projects/plugins/crm/src/automation/class-automation-engine.php

```diff
@@ -179,6 +298,98 @@ public function init_workflows() {
 		}
 	}
 
+	/**
+	 * Execute workflow
+	 *
+	 * @param Automation_Workflow $workflow The workflow to be executed.
+	 * @param Trigger             $trigger The trigger that started the execution process.
+	 * @param array               $trigger_data The data that was passed along by the trigger.
+	 * @return bool
+	 *
+	 * @throws Automation_Exception Throws exception if the step does not exist, or there is an error executing the workflow.
+	 */
+	public function execute_workflow( Automation_Workflow $workflow, Trigger $trigger, array $trigger_data ): bool {
+		$this->get_logger()->log( sprintf( 'Trigger activated: %s', $trigger->get_slug() ) );
+		$this->get_logger()->log( sprintf( 'Executing workflow: %s', $workflow->name ) );
+
+		$step_data = $workflow->get_initial_step();
+		// Convert the trigger data into a data type instance.
+		// This might change while running steps if we have to transform the data
+		// to e.g. allow invoice triggers to work together with contact actions.
+		$hook_data = $this->get_data_type_instance( $trigger::get_data_type(), $trigger_data );
+
+		while ( $step_data ) {
+			try {
+				$step_slug = $step_data['slug'];
+
+				$step_class = $step_data['class_name'] ?? $this->get_step_class( $step_slug );
+
+				if ( ! class_exists( $step_class ) ) {
+					throw new Automation_Exception(
+					/* Translators: %s is the name of the step class that does not exist. */
+						sprintf( __( 'The step class %s does not exist.', 'zero-bs-crm' ), $step_class ),
+						Automation_Exception::STEP_CLASS_NOT_FOUND
+					);
+				}
+
+				/** @var Step $step */
+				$step = new $step_class( $step_data );
+
+				if ( isset( $step_data['attributes'] ) && is_array( $step_data['attributes'] ) ) {
+					$step->set_attributes( $step_data['attributes'] );
+				}
+
+				$this->get_logger()->log( '[' . $step->get_slug() . '] Executing step. Type: ' . $step->get_data_type() );
+
+				$hook_data = $this->maybe_transform_data_type( $hook_data, $step::get_data_type() );
```

Comment:
We should preserve the original data type throughout the workflow and adapt it to the right type at each step.

Reply:
The reason is that a step below in the workflow could need the original type object, for example:

1. On a new invoice created, the trigger receives an invoice object and execute a workflow.
2. Contact action: Set the status:
    - Adapt the input data (Invoice) and pass the contact object to the action.
    - Execute the action: set status to `Customer`
4. Send invoice email action: send to the contact the invoice
    - It needs the invoice.
    - From the invoice, it could get the Contact data.